File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/user/UserTimestamps.java
package top.likoslupus.cellulosesz.api.user;

import org.jspecify.annotations.Nullable;

import static top.likoslupus.cellulosesz.api.validation.Checks.requireNonNegative;

public record UserTimestamps(
        long firstJoin,
        long lastJoin,
        long lastQuit,
        long playTimeMillis,
        long lastActivityAt,
        @Nullable Long activeSessionStartedAt
) {

    public UserTimestamps {
        requireNonNegative(firstJoin, "firstJoin");
        requireNonNegative(lastJoin, "lastJoin");
        requireNonNegative(lastQuit, "lastQuit");
        requireNonNegative(playTimeMillis, "playTimeMillis");
        requireNonNegative(lastActivityAt, "lastActivityAt");
        if (activeSessionStartedAt != null) {
            requireNonNegative(activeSessionStartedAt, "activeSessionStartedAt");
        }
    }

    public static UserTimestamps defaults() {
        return new UserTimestamps(
                0,
                0,
                0,
                0,
                0,
                null
        );
    }

    public UserTimestamps withFirstJoin(long value) {
        return new UserTimestamps(value, lastJoin, lastQuit, playTimeMillis, lastActivityAt, activeSessionStartedAt);
    }

    public UserTimestamps withLastJoin(long value) {
        return new UserTimestamps(firstJoin, value, lastQuit, playTimeMillis, lastActivityAt, activeSessionStartedAt);
    }

    public UserTimestamps withLastQuit(long value) {
        return new UserTimestamps(firstJoin, lastJoin, value, playTimeMillis, lastActivityAt, activeSessionStartedAt);
    }

    public UserTimestamps withPlayTimeMillis(long value) {
        return new UserTimestamps(firstJoin, lastJoin, lastQuit, value, lastActivityAt, activeSessionStartedAt);
    }

    public UserTimestamps withLastActivityAt(long value) {
        return new UserTimestamps(firstJoin, lastJoin, lastQuit, playTimeMillis, value, activeSessionStartedAt);
    }

    public UserTimestamps withActiveSessionStartedAt(@Nullable Long value) {
        return new UserTimestamps(firstJoin, lastJoin, lastQuit, playTimeMillis, lastActivityAt, value);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/user/UserTimestamps.kt
package top.likoslupus.cellulosesz.api.user

import top.likoslupus.cellulosesz.api.validation.requireNonNegative

@JvmRecord
public data class UserTimestamps(
    public val firstJoin: Long,
    public val lastJoin: Long,
    public val lastQuit: Long,
    public val playTimeMillis: Long,
    public val lastActivityAt: Long,
    public val activeSessionStartedAt: Long?
) {

    init {
        firstJoin.requireNonNegative { "firstJoin" }
        lastJoin.requireNonNegative { "lastJoin" }
        lastQuit.requireNonNegative { "lastQuit" }
        playTimeMillis.requireNonNegative { "playTimeMillis" }
        lastActivityAt.requireNonNegative { "lastActivityAt" }
        activeSessionStartedAt?.requireNonNegative { "activeSessionStartedAt" }
    }

    public fun withFirstJoin(value: Long): UserTimestamps =
        UserTimestamps(
            value,
            lastJoin,
            lastQuit,
            playTimeMillis,
            lastActivityAt,
            activeSessionStartedAt
        )

    public fun withLastJoin(value: Long): UserTimestamps =
        UserTimestamps(
            firstJoin,
            value,
            lastQuit,
            playTimeMillis,
            lastActivityAt,
            activeSessionStartedAt
        )

    public fun withLastQuit(value: Long): UserTimestamps =
        UserTimestamps(
            firstJoin,
            lastJoin,
            value,
            playTimeMillis,
            lastActivityAt,
            activeSessionStartedAt
        )

    public fun withPlayTimeMillis(value: Long): UserTimestamps =
        UserTimestamps(
            firstJoin,
            lastJoin,
            lastQuit,
            value,
            lastActivityAt,
            activeSessionStartedAt
        )

    public fun withLastActivityAt(value: Long): UserTimestamps =
        UserTimestamps(
            firstJoin,
            lastJoin,
            lastQuit,
            playTimeMillis,
            value,
            activeSessionStartedAt
        )

    public fun withActiveSessionStartedAt(value: Long?): UserTimestamps =
        UserTimestamps(
            firstJoin,
            lastJoin,
            lastQuit,
            playTimeMillis,
            lastActivityAt,
            value
        )

    public companion object {

        @JvmStatic
        public fun defaults(): UserTimestamps =
            UserTimestamps(
                firstJoin = 0,
                lastJoin = 0,
                lastQuit = 0,
                playTimeMillis = 0,
                lastActivityAt = 0,
                activeSessionStartedAt = null
            )

    }

}
